### PR TITLE
Add appointment controllers

### PR DESCRIPTION
**Pull request by JoshL579**
- change on middleware: able to pass user id into routes after validating token
- add DB relationship between Users and Appointments

TODOs:
- clarify if the appointment name comes from FE or the name in Users
- clarify the workflow of where the eventType is gathered when adding appointments

### CODE REVIEW

File: server/api/apointment_handler.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from config import db
+from utils.auth.middleware import check_token
+import dateutil
+import re
+
+create_appointment_handler = Blueprint('create_appointment_handler', __name__)
+get_appointment_handler = Blueprint('get_appointment_handler', __name__)
+
+
+@create_appointment_handler.route('/appointment/create', methods=['POST'])
```
Comment (fangyu3) [inline server/api/apointment_handler.py]: i think in the FE branch the route was just /appointment

File: server/api/apointment_handler.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    if not re.search(regex, email):
+        return jsonify({'success': False, 'msg': 'Invalid email'}), 400
+
+    try:
+        dateutil.parser.isoparse(time)
+    except:
+        return jsonify({'success': False, 'msg': 'Invalid datetime'}), 400
+
+    user = Users.query.filter_by(id=user_id).first()
+    name = user.name  # need clarify if this name come from FE or the name in DB
```
Comment (fangyu3) [inline server/api/apointment_handler.py]: yep so as discussed during meeting name is actually the booker's name

File: server/api/apointment_handler.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+from flask import jsonify, Blueprint, request
+import json
+from model.model import Users, Appointments
+from config import db
+from utils.auth.middleware import check_token
+import dateutil
+import re
+
+create_appointment_handler = Blueprint('create_appointment_handler', __name__)
+get_appointment_handler = Blueprint('get_appointment_handler', __name__)
```
Comment (fangyu3) [inline server/api/apointment_handler.py]: I think one appointment_handler blue print should be good

File: server/api/apointment_handler.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        time=time,
+        timezone=timezone
+    )
+    db.session.add(appointment)
+    try:
+        db.session.commit()
+    except:
+        return jsonify({'success': False, 'msg': 'Failed to record'}), 500
+    return jsonify({'success': True, 'msg': 'Appointment Booked'}), 200
+
```
Comment (fangyu3) [inline server/api/apointment_handler.py:71]: We also need to sync with Google Calendar. For this i think we need to get the duration of event and do something like this: https://developers.google.com/calendar/create-events

make sure to put function under utils/auth/google_client for better organization!

File: server/api/apointment_handler.py (hunk 1 above)
Comment (JoshL579) [reply server/api/apointment_handler.py]: oh, right. I'll change the FE later

File: server/api/apointment_handler.py (hunk 4 above)
Comment (JoshL579) [reply server/api/apointment_handler.py:71]: Ok, I'll try it later today

Comment (JoshL579) [reply server/api/apointment_handler.py:71]: and also add the sendemail

File: server/api/google_login_handler.py [hunk 5]
```diff
@@ -20,6 +21,9 @@ def googlelogin():
             user_db = Users.query.filter_by(google_id=userid).first()
             uid = user_db.id
             token = token_generator(uid)
+            user_db.access_token = user_info.get('access_token')
+            user_db.refresh_token = user_info.get('refresh_token')
+            db.session.commit()
```
Comment (JoshL579) [inline server/api/google_login_handler.py:27]: @fangyu3 I've found that the `access_token` and `refresh_token` are not recorded when user `login`, it is only recorded when the user `signs up`. But when I test with the initial tokens made when I signed up, the calendar event creation doesn't recognize my credential. So I added this in `login` don't know if it is necessary but it does solved my problem.

File: server/utils/auth/google_client.py [hunk 6]
```diff
@@ -85,4 +87,37 @@ def get_user_calendar_info(self, timeMin, timeMax, timezone):
...
+                'useDefault': False,
+                'overrides': [
+                    {'method': 'email', 'minutes': 24 * 60},
+                    {'method': 'popup', 'minutes': 10},
+                ],
+            },
+        }
+
+        event_service = build('calendar', 'v3', credentials=self.credentials)
+        event = event_service.events().insert(calendarId='primary', body=event).execute()
```
Comment (JoshL579) [inline server/utils/auth/google_client.py:122]: @fangyu3 and please review this part, if no problem I'll merge then

File: server/api/apointment_handler.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        db.session.commit()
+    except:
+        return jsonify({'success': False, 'msg': 'Failed to record'}), 500
+
+    return jsonify({'success': True, 'msg': 'Appointment Booked'}), 200
+
+
+@appointment_handler.route('/appointment/<uid>', methods=['GET'])
+@check_token
+def get_appointment(uid):
```
Comment (fangyu3) [inline server/api/apointment_handler.py:75]: For this function, is the uid the user's id? or event type id?

Judging by the query (event type id === uid) it seems it is getting all appointments with the same event type? But i feel we should be getting all appointments of a user. Unless you meant event type.user_id === uid?

File: server/api/google_login_handler.py (hunk 5 above)
Comment (fangyu3) [reply server/api/google_login_handler.py:27]: hmm i would think the tokens from sign up is enough, but i guess not. As long as it solves the issue, it's a good solution.

File: server/utils/auth/google_client.py (hunk 6 above)
Comment (fangyu3) [reply server/utils/auth/google_client.py:122]: yep i see no issues! If you tested it and it works then it's good to go!

File: server/api/apointment_handler.py (hunk 4 above)
Comment (fangyu3) [reply server/api/apointment_handler.py:71]: Did you get a chance to add the send email part?

Comment (JoshL579) [reply server/api/apointment_handler.py:71]: No, cuz I realized that the email could be sent from google once the event created. So i was thing if we need sendemail or not. And if we need, i can do it in the next branch, the intergration

File: server/api/apointment_handler.py (hunk 7 above)
Comment (JoshL579) [reply server/api/apointment_handler.py:75]: Oh, good catch, that should be eventTypes.user_id ==uid.
Dose it make sense?

Comment (fangyu3) [reply server/api/apointment_handler.py:75]: yep eventTypes.user_id ==uid makes sense

File: server/api/apointment_handler.py (hunk 4 above)
Comment (fangyu3) [reply server/api/apointment_handler.py:71]: okk got it

**Review by fangyu3 — APPROVED** [review]

**Pull request merged**